### PR TITLE
A4A: Add "Next Steps" Onboarding Tours

### PR DESCRIPTION
**Pull request by nateweller**
Resolves https://github.com/Automattic/jetpack-genesis/issues/292

## Proposed Changes

* Adds the guided tour components to the Sites screen.
  * The tour will be shown when the page's query string includes `tour=sites-walkthrough` or `tour=add-new-site`. These are linked in the Overview page's "Next Steps" card.
  * You can click step by step through each popover.
  * If you open a site details pane half way through a tour, the popover component will be hidden automatically. Closing the site will resume the tour.
  * Uses React Context to provide data about the current tour to all `GuidedTourStep` components. 
  * Uses individual `GuidedTourStep` components to ensure each step is only rendered when it's anchor is available.

## Testing Instructions

* From the A4A overview page, click the "get familiar with the sites management dashboard" and "learn how to add new sites" steps to start each tour.
* Verify you are redirected to the sites screen with a `tour` parameter in the query string.
* Click through the tour and verify each step renders appropriately.
* Verify skipping the tour dismisses all steps.
* Verify a dismissed tour does not re-start on page refresh.

## Pre-merge Checklist

- [x] Has the general commit checklist been followed? (PCYsg-hS-p2)
- [ ] [Have you written new tests](https://wpcalypso.wordpress.com/devdocs/docs/testing/index.md) for your changes?
- [ ] Have you tested the feature in Simple (P9HQHe-k8-p2), Atomic (P9HQHe-jW-p2), and self-hosted Jetpack sites (PCYsg-g6b-p2)?
- [x] Have you checked for TypeScript, React or other console errors?
- [x] Have you used memoizing on expensive computations? More info in [Memoizing with create-selector](https://github.com/Automattic/wp-calypso/blob/trunk/packages/state-utils/src/create-selector/README.md) and [Using memoizing selectors](https://react-redux.js.org/api/hooks#using-memoizing-selectors) and [Our Approach to Data](https://github.com/Automattic/wp-calypso/blob/trunk/docs/our-approach-to-data.md)
- [ ] Have we added the "[Status] String Freeze" label as soon as any new strings were ready for translation (p4TIVU-5Jq-p2)?
- [ ] For changes affecting Jetpack: Have we added the "[Status] Needs Privacy Updates" label if this pull request changes what data or activity we track or use (p4TIVU-aUh-p2)?

## Screenshots

Screenshot 2024-04-01 at 9 15 20 PM

### CODE REVIEW

File: client/a8c-for-agencies/components/guided-tour-step/style.scss [hunk 1]
```diff
@@ -0,0 +1,49 @@
...
+		padding: 16px;
+		flex-direction: column;
+		align-items: flex-start;
+		/* stylelint-disable-next-line scales/radii */
+		border-radius: 8px;
+	}
+}
+
+.guided-tour__popover-heading {
+	color: var(--studio-gray-100);
```
Comment (yashwin) [inline client/a8c-for-agencies/components/guided-tour-step/style.scss]: It is recommended (https://wpcalypso.wordpress.com/devdocs/docs/color.md) to avoid using `--studio` variables directly favoring the Theme properties. (e.g. --color-neutral-100). This allows us to easily override color schemes without modifying the CSS code.

File: client/a8c-for-agencies/sections/sites/sites-dataviews/style.scss [hunk 2]
```diff
@@ -0,0 +1,215 @@
...
+
+.dataviews-pagination {
+	margin-bottom: 2rem;
+}
+
+
+.dataviews-wrapper {
+	.dataviews-no-results,
+	.dataviews-loading {
+		padding-top: 1rem;
```
Comment (yashwin) [inline client/a8c-for-agencies/sections/sites/sites-dataviews/style.scss:188]: [Non-blocker] It is better to use [logical properties](https://developer.mozilla.org/en-US/docs/Web/CSS/CSS_logical_properties_and_values) everywhere on this file since we can support RTL locales in Calypso.

File: client/a8c-for-agencies/sections/sites/sites-dataviews/style.scss [hunk 3]
```diff
@@ -0,0 +1,215 @@
+@import "@wordpress/base-styles/breakpoints";
+@import "@wordpress/base-styles/mixins";
+
+.sites-overview__content {
+
+	.components-checkbox-control__input[type="checkbox"]:focus {
+		box-shadow: 0 0 0 var(--studio-jetpack-green-50) #fff, 0 0 0 calc(2* var(--studio-jetpack-green-50)) var(--studio-jetpack-green-50);
```
Comment (yashwin) [inline client/a8c-for-agencies/sections/sites/sites-dataviews/style.scss:7]: It is better to use the theme properties. (e.g. --color-neutral-100).

Also, I think we should use the A4A theme color instead of jetpack-green here.

**Review by yashwin — COMMENTED** [review]
Good job on this, @nateweller!

It mostly looks good. I have left a few comments. Let me know what you think.

Few things to note:

- Could you please rebase this PR with the latest trunk?
- I also see the PR has 1300+ lines added. [We suggest keeping it to ](pedEI5-116-p2) maximum of 500 lines, limiting the scope of changes as tightly as makes sense, and breaking it into multiple PRs, as it becomes very hard to review everything. But it is okay for now as I have reviewed most of the things.

File: client/a8c-for-agencies/sections/sites/sites-dataviews/style.scss (hunk 3 above)
Comment (cleacos) [reply client/a8c-for-agencies/sections/sites/sites-dataviews/style.scss:7]: One note about the style from sites-dataviews. This component is a copy from Manage. As a task, we have to audit the whole style in A4A and change it by the theme color. And once the new [Automattic color palette](https://github.com/Automattic/color-studio/pull/633#top) is added to the Studio package, we can update the theme color to those studio ones: for instance: `--color-primary: var( --studio-automattic-blue-50 );`

File: client/a8c-for-agencies/components/guided-tour-step/style.scss (hunk 1 above)
Comment (nateweller) [reply client/a8c-for-agencies/components/guided-tour-step/style.scss]: Fixed via [e7e7a8d](https://github.com/Automattic/wp-calypso/pull/89081/commits/e7e7a8d646f769f740053d389ad7a9657a8727a4)

**Comment by nateweller** [comment]
Thanks for the review @yashwin! I've rebased trunk and addressed the color variables. I've left the sites dataviews alone for now. Let me know what you think :+1:

**Review by andrii-lysenko — DISMISSED** [review]
<del>Haven't tested it manually but the code looks good and it already had previous approvals! 👍 </del>
It was a comment for another PR.

**Comment by cleacos** [comment]
There is a failing test: https://teamcity.a8c.com/buildConfiguration/calypso_calypso_WebApp_Run_All_Unit_Tests/12230574?expandBuildDeploymentsSection=false&hideTestsFromDependencies=false&hideProblemsFromDependencies=false&expandBuildProblemsSection=true&showLog=12230574_10582_532.10460.10581.10608&logView=flowAware

File: client/a8c-for-agencies/components/layout/index.tsx [hunk 4]
```diff
@@ -1,5 +1,7 @@
 import classNames from 'classnames';
 import React, { ReactNode } from 'react';
+import GuidedTour from 'calypso/a8c-for-agencies/components/guided-tour';
```
Comment (andrii-lysenko) [inline client/a8c-for-agencies/components/layout/index.tsx:3]: I think this variable is unused.

File: client/a8c-for-agencies/components/guided-tour-step/index.tsx [hunk 5]
```diff
@@ -0,0 +1,133 @@
...
+import { savePreference } from 'calypso/state/preferences/actions';
+import {
+	getPreference,
+	preferencesLastFetchedTimestamp,
+} from 'calypso/state/preferences/selectors';
+
+type Props = {
+	id: string;
+	tourId: TourId;
+	context?: LegacyRef< HTMLElement | null >;
```
Comment (andrii-lysenko) [inline client/a8c-for-agencies/components/guided-tour-step/index.tsx]: There's a typescript error caused by this. I think we can use just `HTMLElement | null` here.

File: client/a8c-for-agencies/sections/sites/site-preview-pane/index.tsx [hunk 6]
```diff
@@ -86,11 +88,18 @@ export default function SitePreviewPane( {
...
+			<div ref={ setNavRef }>
+				<SectionNav className="preview-pane__navigation" selectedText={ selectedFeature.tab.label }>
+					{ navItems && navItems.length > 0 && canDisplayNavTabs ? (
+						<NavTabs>{ navItems }</NavTabs>
+					) : null }
+				</SectionNav>
+			</div>
+			<GuidedTourStep
+				id="sites-walkthrough-site-preview-tabs"
+				tourId="sitesWalkthrough"
```
Comment (andrii-lysenko) [inline client/a8c-for-agencies/sections/sites/site-preview-pane/index.tsx:100]: I see typescript shows an error here. Should it be `sites-walkthrough`?

File: client/a8c-for-agencies/sections/sites/sites-dataviews/index.tsx [hunk 7]
```diff
@@ -0,0 +1,474 @@
...
+						siteError={ item.site.error }
+					/>
+				);
+			}
+		},
+		[ isLoading, isLargeScreen ]
+	);
+
+	// Legacy refs for guided tour popovers
+	const [ introRef, setIntroRef ] = useState< React.LegacyRef< HTMLElement | null > >();
```
Comment (andrii-lysenko) [inline client/a8c-for-agencies/sections/sites/sites-dataviews/index.tsx]: Here we should be able to use simply as well
```
const [ introRef, setIntroRef ] = useState< HTMLElement | null >();
```

File: client/a8c-for-agencies/sections/sites/sites-dataviews/index.tsx [hunk 8]
```diff
@@ -0,0 +1,474 @@
...
+							<span
+								className="sites-dataview__site-header sites-dataview__site-header--sort"
+								ref={ ( ref ) => setIntroRef( ref as React.LegacyRef< HTMLElement | null > ) }
+							>
+								{ translate( 'Site' ).toUpperCase() }
+							</span>
+						</SiteSort>
+						<GuidedTourStep
+							id="sites-walkthrough-intro"
+							tourId="sitesWalkthrough"
```
Comment (andrii-lysenko) [inline client/a8c-for-agencies/sections/sites/sites-dataviews/index.tsx:121]: Typescript is complaining here. Should it be `tourId="sites-walkthrough"`?

File: client/a8c-for-agencies/sections/sites/sites-header-actions/index.tsx [hunk 9]
```diff
@@ -0,0 +1,51 @@
...
+					}
+					onClickJetpackMenuItem={ () =>
+						dispatch( recordTracksEvent( 'calypso_a4a_sites_connect_jetpack_site_click' ) )
+					}
+					onClickUrlMenuItem={ () =>
+						dispatch( recordTracksEvent( 'calypso_a4a_sites_connect_url_site_click' ) )
+					}
+				/>
+			</div>
+			<GuidedTourStep id="add-new-site" tourId="addSiteStep1" context={ tourStepRef } />
```
Comment (andrii-lysenko) [inline client/a8c-for-agencies/sections/sites/sites-header-actions/index.tsx:39]: Should `tourId` here be `'add-new-site'`?

File: client/a8c-for-agencies/components/guided-tour-step/index.tsx [hunk 10]
```diff
@@ -0,0 +1,133 @@
...
+								{ translate( 'Step %(currentStep)d of %(totalSteps)d', {
+									args: { currentStep: currentStepCount, totalSteps: stepsCount },
+								} ) }
+							</span>
+						)
+					}
+				</div>
+				<div className="guided-tour__popover-footer-right-content">
+					<>
+						{ ( ( ! currentStep.nextStepOnTargetClick &&
```
Comment (andrii-lysenko) [inline client/a8c-for-agencies/components/guided-tour-step/index.tsx:116]: It's a bit hard to follow here. Maybe we could move it to a constant defined above?

File: client/a8c-for-agencies/sections/overview/body/next-steps/index.tsx [hunk 11]
```diff
@@ -37,9 +59,7 @@ export default function OverviewBodyNextSteps() {
 			disabled: false,
 			actionDispatch: () => {
 				dispatch( recordTracksEvent( 'calypso_a4a_overview_next_steps_get_familiar_click' ) );
-				dispatch(
-					savePreference( A4A_ONBOARDING_TOURS_PREFERENCE_NAME[ 'sitesWalkthrough' ], true )
-				);
+				resetTour( [ 'sitesWalkthrough' ] );
```
Comment (andrii-lysenko) [inline client/a8c-for-agencies/sections/overview/body/next-steps/index.tsx:62]: should it be `sites-walkthrough`?

**Review by andrii-lysenko — CHANGES_REQUESTED** [review]
Thank you for working on this. Overall, it looks like a solid solution.

There's a number of small issues with typescript build we need to address prior to be able to fully test it though. Please let me know if these looks reasonable. Thanks!

File: client/a8c-for-agencies/components/guided-tour-step/index.tsx [hunk 12]
```diff
@@ -71,13 +71,13 @@ export function GuidedTourStep( { id, tourId, context, hideSteps, className }: P
 	// Add a click event listener to the context element if the step requires it
 	useEffect( () => {
 		if ( currentStep && currentStep.nextStepOnTargetClick && context ) {
-			context.addEventListener( 'click', nextStep );
+			( context as HTMLElement ).addEventListener( 'click', nextStep );
```
Comment (andrii-lysenko) [inline client/a8c-for-agencies/components/guided-tour-step/index.tsx:76]: I don't think it's necessary here. Should be ok just to use `context.addEventListener`?

File: client/a8c-for-agencies/sections/sites/sites-dataviews/index.tsx [hunk 13]
```diff
@@ -111,14 +111,14 @@ const SitesDataViews = ( {
 						<SiteSort isSortable={ true } columnKey="site">
 							<span
 								className="sites-dataview__site-header sites-dataview__site-header--sort"
-								ref={ ( ref ) => setIntroRef( ref as React.LegacyRef< HTMLElement | null > ) }
+								ref={ ( ref ) => setIntroRef( ref as HTMLElement | null ) }
```
Comment (andrii-lysenko) [inline client/a8c-for-agencies/sections/sites/sites-dataviews/index.tsx:114]: Similarly, no need for type assertion here.

**Review by andrii-lysenko — APPROVED** [review]
Great work 👍  Thank you for addressing the comments. Looks great now. Let's merge it and iterate if needed.

Some small comments, that we don't need type assertions but they are not blocking.
